core/src/main/java/jenkins/security/ResourceDomainRootAction.java

```diff
@@ -0,0 +1,278 @@
+
+    @CheckForNull
+    @Override
+    public String getIconFileName() {
+        return null;
+    }
+
+    @CheckForNull
+    @Override
+    public String getDisplayName() {
+        return null;
+    }
+
+    @CheckForNull
+    @Override
+    public String getUrlName() {
+        return "static-files";
+    }
+
+    public void doIndex(StaplerRequest req, StaplerResponse rsp) throws IOException {
+        if (!ResourceDomainConfiguration.isResourceRequest(req)) {
+            rsp.sendError(404, "Cannot handle requests to this URL unless on Jenkins resource URL.");
+        } else {
+            rsp.sendError(404, ResourceDomainFilter.ERROR_RESPONSE);
+        }
+    }
+
+    public Object getDynamic(String id, StaplerRequest req, StaplerResponse rsp) throws Exception {
+        if (!ResourceDomainConfiguration.isResourceRequest(req)) {
+            rsp.sendError(404, "Cannot handle requests to this URL unless on Jenkins resource URL.");
+            return null;
+        }
+
+        Token token = Token.decode(id);
+        if (token == null) {
+            rsp.sendError(404, ResourceDomainFilter.ERROR_RESPONSE);
+            return null;
+        }
+
+        String authenticationName = token.username;
+        String browserUrl = token.path;
+
+
+        Instant creationDate = ofEpochMilli(token.timestamp);
+        if (creationDate.plus(VALID_FOR_MINUTES, MINUTES).isAfter(now()) && creationDate.isBefore(now())) {
+            return new InternalResourceRequest(browserUrl, authenticationName);
+        }
+
+        // too old, so redirect to the real file first
+        return new Redirection(browserUrl);
+    }
+
+    private static class Redirection {
+        private final String url;
+
+        public Redirection(String url) {
+            this.url = url;
+        }
+
+        public void doDynamic(StaplerRequest req, StaplerResponse rsp) throws IOException {
+            String restOfPath = req.getRestOfPath();
+
+            String url = Jenkins.get().getRootUrl() + this.url + restOfPath;
+            rsp.sendRedirect(302, url);
+        }
+    }
+
+    public String getRedirectUrl(Token token, String restOfPath) {
+        String resourceRootUrl = getResourceRootUrl();
+        if (!resourceRootUrl.endsWith("/")) {
+            resourceRootUrl += "/";
+        }
+        if (!restOfPath.startsWith("/")) {
+            // Unsure whether this can happen -- just be safe here
+            restOfPath = "/" + restOfPath;
+        }
+        return resourceRootUrl + getUrlName() + "/" + token.encode() + restOfPath;
+    }
+
+    private static String getResourceRootUrl() {
+        return ResourceDomainConfiguration.get().getUrl();
+    }
+
+    public Token getToken(DirectoryBrowserSupport dbs, StaplerRequest req) {
+        String dbsFile = req.getRestOfPath();
+
+        String completeUrl = req.getAncestors().get(0).getRestOfUrl();
+        completeUrl = completeUrl.substring(0, completeUrl.length() - dbsFile.length());
```

Comment:
This stuff is sufficiently weird that it would be polite to give some example values in comments.